File: frontend/src/types/securityPatrol.ts
/**
 * 保全巡檢 TypeScript 型別定義
 */

// ── Sheet 設定 ────────────────────────────────────────────────────────────────
export interface SheetConfig {
  key:  string
  id:   number
  name: string
  path: string
}

// ── 巡檢場次 ──────────────────────────────────────────────────────────────────
export interface PatrolBatch {
  ragic_id:        string
  sheet_key:       string
  sheet_id:        number
  sheet_name:      string
  inspection_date: string
  inspector_name:  string
  start_time:      string
  end_time:        string
  work_hours:      string
  item_count:      number
  synced_at?:      string | null
}

// ── 巡檢項目 ──────────────────────────────────────────────────────────────────
export interface PatrolItem {
  ragic_id:       string
  batch_ragic_id: string
  sheet_key:      string
  seq_no:         number
  item_name:      string
  result_raw:     string
  result_status:  'normal' | 'abnormal' | 'pending' | 'unchecked' | 'note'
  abnormal_flag:  boolean
  is_note?:       boolean  // true = 文字備註欄位（異常說明），不計入統計
  synced_at?:     string | null
}

// ── KPI ───────────────────────────────────────────────────────────────────────
export interface PatrolBatchKPI {
  total:           number
  normal:          number
  abnormal:        number
  pending:         number
  unchecked:       number
  completion_rate: number
  normal_rate:     number
}

export interface StatusDistItem {
  status: string
  label:  string
  count:  number
  color:  string
}

// ── 批次詳情 ──────────────────────────────────────────────────────────────────
export interface PatrolBatchDetail {
  batch: PatrolBatch
  kpi:   PatrolBatchKPI
  items: PatrolItem[]
}

// ── 場次清單項目（含 KPI）────────────────────────────────────────────────────
export interface PatrolBatchListItem {
  batch: PatrolBatch
  kpi:   PatrolBatchKPI
}

// ── 全站統計 ──────────────────────────────────────────────────────────────────
export interface PatrolStats {
  sheet_key:           string
  sheet_name:          string
  latest_batch?:       PatrolBatch | null
  latest_kpi?:         PatrolBatchKPI | null
  recent_abnormal:     PatrolItem[]
  recent_pending:      PatrolItem[]
  status_distribution: StatusDistItem[]
  total_batches_7d:    number
  abnormal_trend:      AbnormalTrendPoint[]
}

export interface AbnormalTrendPoint {
  date:           string
  abnormal_count: number
  has_record:     boolean
}

// ── Dashboard 型別 ────────────────────────────────────────────────────────────
export interface SheetStats {
  sheet_key:       string
  sheet_name:      string
  total_batches:   number
  total_items:     number
  checked_items:   number
  unchecked_items: number
  abnormal_items:  number
  pending_items:   number
  completion_rate: number
  normal_rate:     number
  has_data:        boolean
}

export interface SecurityDashboardSummary {
  target_date:           string
  sheets:                SheetStats[]
  total_batches_all:     number
  total_items_all:       number
  checked_items_all:     number
  abnormal_items_all:    number
  completion_rate_all:   number
  generated_at:          string
}

export interface SecurityIssueItem {
  id:           string
  issue_date:   string
  sheet_key:    string
  sheet_name:   string
  item_name:    string
  status:       string
  status_label: string
  inspector:    string
  note:         string
  batch_id:     string
}

export interface SecurityIssueListResponse {
  items: SecurityIssueItem[]
  total: number
}

export interface SecurityTrendPoint {
  date:           string
  abnormal_count: number
  total_batches:  number
  has_data:       boolean
  by_sheet:       Record<string, { batch_count: number; abnormal: number }>
}

export interface SecurityDashboardTrend {
  trend: SecurityTrendPoint[]
  days:  number
}

File: backend/app/schemas/security_patrol.py
"""
保全巡檢 Pydantic Schemas
"""
from __future__ import annotations
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel


# ── 場次 ──────────────────────────────────────────────────────────────────────

class PatrolBatchOut(BaseModel):
    ragic_id:        str
    sheet_key:       str
    sheet_id:        int
    sheet_name:      str
    inspection_date: str
    inspector_name:  str
    start_time:      str
    end_time:        str
    work_hours:      str
    item_count:      int = 0
    synced_at:       Optional[datetime] = None

    class Config:
        from_attributes = True


# ── 巡檢項目 ──────────────────────────────────────────────────────────────────

class PatrolItemOut(BaseModel):
    ragic_id:       str
    batch_ragic_id: str
    sheet_key:      str
    seq_no:         int
    item_name:      str
    result_raw:     str
    result_status:  str   # normal / abnormal / pending / unchecked / note
    abnormal_flag:  bool
    is_note:        bool = False  # True = 文字備註欄位（異常說明），呈現但不計入統計
    synced_at:      Optional[datetime] = None

    class Config:
        from_attributes = True


# ── KPI ───────────────────────────────────────────────────────────────────────

class PatrolBatchKPI(BaseModel):
    total:           int   = 0
    normal:          int   = 0
    abnormal:        int   = 0
    pending:         int   = 0
    unchecked:       int   = 0
    completion_rate: float = 0.0
    normal_rate:     float = 0.0


class StatusDistItem(BaseModel):
    status: str
    label:  str
    count:  int
    color:  str


# ── 批次詳情 ──────────────────────────────────────────────────────────────────

class PatrolBatchDetail(BaseModel):
    batch: PatrolBatchOut
    kpi:   PatrolBatchKPI
    items: List[PatrolItemOut]


# ── 全站統計（Dashboard 資料來源）────────────────────────────────────────────

class PatrolStats(BaseModel):
    sheet_key:          str
    sheet_name:         str
    latest_batch:       Optional[PatrolBatchOut] = None
    latest_kpi:         Optional[PatrolBatchKPI] = None
    recent_abnormal:    List[PatrolItemOut]       = []
    recent_pending:     List[PatrolItemOut]       = []
    status_distribution: List[StatusDistItem]     = []
    total_batches_7d:   int                       = 0
    abnormal_trend:     List[dict]                = []


# ── Sheet 設定資訊 ─────────────────────────────────────────────────────────────

class SheetConfig(BaseModel):
    key:  str
    id:   int
    name: str
    path: str
